### PR TITLE
Snake tutorial

### PR DESCRIPTION
**Pull request by christopher-kleine**
Started a tutorial about how to create a snake using WASM-4.
This is a Work-In-Progress and should NOT be merged for now.

The purpose of this PR is so that others can stop me from doing something realy stupid and improve it along the way.

### CODE REVIEW

File: site/docs/tutorials/snake/moving-the-snake.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
+---
+sidebar_label: Moving the snake
```
Comment (aduros) [inline site/docs/tutorials/snake/moving-the-snake.md:2]: Is `sidebar_label` needed? I think docusaurus should be able to infer it from the first title in the document?

File: site/docs/tutorials/snake/moving-the-snake.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    groupId="code-language"
+    defaultValue="language-typescript"
+    values={[
+        {label: 'AssemblyScript', value: 'language-typescript'},
+        {label: 'C / C++', value: 'language-cpp'},
+        {label: 'Rust', value: 'language-rust'},
+        {label: 'Go', value: 'language-go'},
+    ]}>
+
+<TabItem value="language-typescript">
```
Comment (aduros) [inline site/docs/tutorials/snake/moving-the-snake.md:39]: Is the idea that these other languages will be filled out later? Right now it looks a bit broken when Go is not selected.

Maybe we should remove the language pickers for now, and explain that this tutorial is geared towards Go?

File: site/docs/tutorials/snake/goal.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+- ... draw primitives (rectangles)
+- ... convert a sprite-image to source and draw it
+- ... handle user-input
+- ... publish your game
+
+## Preview
+
+At the end of this tutorial, your snake game will look like this:
+
+![Snake Preview](images/endresult.webp)
```
Comment (aduros) [inline site/docs/tutorials/snake/goal.md:27]: I really like this page :smile: It's nice to have an overview of what will be built and take a peek at the final result.

File: site/docs/tutorials/snake/handling-user-input.md [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+This is true for all other buttons too.
+
+## Keyboard-Layout
+
+For the player side of things, WASM-4 tries to cover most keyboard-layouts:
+
+- X and Space = Button 1
+- Y, C and Z = Button 2
+
+This should cover QWERTY, QWERTZ and Dovrak layouts.
```
Comment (aduros) [inline site/docs/tutorials/snake/handling-user-input.md:42]:
```suggestion
This should cover QWERTY, QWERTZ and Dvorak layouts.
```

File: site/docs/tutorials/snake/collision-detection.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
+---
+sidebar_label: Collision detection
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Collision detection ...
```
Comment (aduros) [inline site/docs/tutorials/snake/collision-detection.md:8]:
```suggestion
# Collision Detection
```

File: site/docs/tutorials/snake/collision-detection.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+Your game is progressing nicely. Only two more things and you're done: Growing the snake and "Game Over". For the first, you need to check if the snake collides with the fruit. For the second, you need to check if the snake collides with itself.
+
+## ... with the fruit
```
Comment (aduros) [inline site/docs/tutorials/snake/collision-detection.md:12]:
```suggestion
## Collision Detection with the Fruit
```

File: site/docs/tutorials/snake/collision-detection.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	}
+```
+
+Now you're almost done. Only "Game Over" is left to finish this game.
+
+</TabItem>
+
+</Tabs>
+
+## ... with itself
```
Comment (aduros) [inline site/docs/tutorials/snake/collision-detection.md:76]:
```suggestion
## Collision Detection with Itself
```

The "..." is a little less readable imo, especially this far down the page where the reader may have forgotten what it was connected to.

File: site/docs/tutorials/snake/creating-the-snake.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
+---
+sidebar_label: Creating the snake
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Creating the snake
```
Comment (aduros) [inline site/docs/tutorials/snake/creating-the-snake.md:8]:
```suggestion
# Creating the Snake
```

File: site/docs/tutorials/snake/drawing-the-snake.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
+---
+sidebar_label: Drawing the snake
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Drawing the snake
```
Comment (aduros) [inline site/docs/tutorials/snake/drawing-the-snake.md:8]:
```suggestion
# Drawing the Snake
```

File: site/docs/tutorials/snake/drawing-the-snake.md [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+## Drawing the body
```
Comment (aduros) [inline site/docs/tutorials/snake/drawing-the-snake.md:10]:
```suggestion
## Drawing the Body
```

File: site/docs/tutorials/snake/drawing-the-snake.md [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+You should see some green blocks at the top. 
+
+![Snake Body](images/draw-body.webp)
+
+</TabItem>
+
+</Tabs>
+
+## Drawing the head
```
Comment (aduros) [inline site/docs/tutorials/snake/drawing-the-snake.md:105]:
```suggestion
## Drawing the Head
```

File: site/docs/tutorials/snake/further-ideas.md [hunk 12]
```diff
@@ -0,0 +1,9 @@
+---
+sidebar_label: Further ideas
+---
+
+# Further ideas
```
Comment (aduros) [inline site/docs/tutorials/snake/further-ideas.md:5]:
```suggestion
# Further Ideas
```

File: site/docs/tutorials/snake/handling-user-input.md [hunk 13]
```diff
@@ -0,0 +1,330 @@
+---
+sidebar_label: Handling user input
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Handling user input
```
Comment (aduros) [inline site/docs/tutorials/snake/handling-user-input.md:8]:
```suggestion
# Handling User Input
```

File: site/docs/tutorials/snake/handling-user-input.md [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+
+With a moving snake, you've done a good part of the game. This part let's you give the control into the players hands. If the player presses a certain button, you just have to change the values of the "Direction"-Property of your snake.
+
+That's it.
+
+But first, you need to understand how WASM-4 handles it's user input.
+
+## Gamepad basics
```
Comment (aduros) [inline site/docs/tutorials/snake/handling-user-input.md:16]:
```suggestion
## Gamepad Basics
```

File: site/docs/tutorials/snake/handling-user-input.md [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+
+So if the player presses "Right" and "Button 1", the value would be 33. Now are all values "a power of two", meaning you can set and check them using binary operators.
+
+If you wnt to check if Button 1 is pressed, simply use the binary AND:
+
+This is true for all other buttons too.
+
+## Keyboard-Layout
+
+For the player side of things, WASM-4 tries to cover most keyboard-layouts:
```
Comment (aduros) [inline site/docs/tutorials/snake/handling-user-input.md:37]:
```suggestion
For the player side of things, WASM-4 tries to cover most keyboard layouts:
```

File: site/docs/tutorials/snake/moving-the-snake.md [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+
+Now if you execute this, you'd notice that you can't see much. In fact, you might see the snake for a short moment before the head is all that's left.
+
+![](images/snake_move_head_only.webp)
+
+</TabItem>
+
+</Tabs>
+
+## Moving the head
```
Comment (aduros) [inline site/docs/tutorials/snake/moving-the-snake.md:77]:
```suggestion
## Moving the Head
```

File: site/docs/tutorials/snake/placing-the-fruit.md [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+	frameCount = 0
+	prevState  uint8
++	fruit      image.Point
+```
+
+</TabItem>
+
+</Tabs>
+
+## Random numbers
```
Comment (aduros) [inline site/docs/tutorials/snake/placing-the-fruit.md:45]:
```suggestion
## Random Numbers
```

File: site/docs/tutorials/snake/placing-the-fruit.md [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
++	fruitSprite = [16]byte{0x00, 0xa0, 0x02, 0x00, 0x0e, 0xf0, 0x36, 0x5c, 0xd6, 0x57, 0xd5, 0x57, 0x35, 0x5c, 0x0f, 0xf0}
+```
+
+With that out of the way, it's time to actually render the newly imported sprite.
+
+</TabItem>
+
+</Tabs>
+
+## Rendering a PNG file
```
Comment (aduros) [inline site/docs/tutorials/snake/placing-the-fruit.md:183]:
```suggestion
## Rendering a PNG File
```

File: site/docs/tutorials/snake/publish-your-game.md [hunk 19]
```diff
@@ -0,0 +1,127 @@
+---
+sidebar_label: Publish your game
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Publish your game
+
+## Create a publish version
```
Comment (aduros) [inline site/docs/tutorials/snake/publish-your-game.md:10]:
```suggestion
## Create a Publish Version
```

File: site/docs/tutorials/snake/publish-your-game.md [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+
+:::note Early version
+The WASM-4 project is still in it's early stages. The process of submitting your game onto the website is still not finished and a little rough around the edges.
+:::
+
+WASM-4 and it's website are stored in a Git-Repo (short for Repository) hosted on GitHub.
+
+To publish your game, you *need* a GitHub Account. If you don't have one already, [create one](https://github.com/join). In case you already have one, [login](https://github.com/login).
+
+## Forking the repo
```
Comment (aduros) [inline site/docs/tutorials/snake/publish-your-game.md:80]:
```suggestion
## Forking the Repo
```

File: site/docs/tutorials/snake/publish-your-game.md [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+
+Now you need to "fork" the repo. For this, go to the [original GitHub Repo of WASM-4](https://github.com/aduros/wasm4).
+
+At the top you will find some "actions" you can take: `Watch/Unwatch`, `Star/Unstar`, `Fork`. The last one is what you're looking for.
+
+## Uploading your game
```
Comment (aduros) [inline site/docs/tutorials/snake/publish-your-game.md:86]:
```suggestion
## Uploading Your Game
```

File: site/docs/tutorials/snake/publish-your-game.md [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+
+Once you've forked the repo, visit your version and navigate to the folder `site/static/carts/`. There you can upload your cart.wasm and a 160x160 cover-image of the game. A screenshot is an easy way. Simply press F9 while your game is running.
+
+Also: Rename your cart.wasm to something meaningful. Like `snake-spartan-edition.wasm`. Your screenshot should have the same name (but a different extension like `snake-spartan-edition.png`).
+
+Commit the changes and you're done with this folder.
+
+## Changing the list of games
```
Comment (aduros) [inline site/docs/tutorials/snake/publish-your-game.md:94]:
```suggestion
## Changing the List of Games
```

File: site/docs/tutorials/snake/publish-your-game.md [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+| Property | Description |
+|----------|-------------|
+| slug     | Displayed in the address bar. A `lakeshooter` will result in this: https://wasm4.org/play/lakeshooter a `pong` would result in https://wasm4.org/play/pong |
+| title    | This is the name displayed on the collection. |
+| author   | The name displayed by WASM-4.  |
+| github   | Your GitHub username. This is used to display your avatar on the cover-page. |
+
+Once you've done this, commit those changes too.
+
+## Letting WASM-4 know
```
Comment (aduros) [inline site/docs/tutorials/snake/publish-your-game.md:111]:
```suggestion
## Letting WASM-4 Know
```

File: site/docs/tutorials/snake/publish-your-game.md [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+
+Your fork is now ready. Switch to the `Pull Requests` tab and press on "New pull request".
+
+![Pull Request](images/pull-request.png)
+
+Add some comment if you want. And then press "Make pull request" at the bottom. You're done!
+
+## Wait a bit
```
Comment (aduros) [inline site/docs/tutorials/snake/publish-your-game.md:119]:
```suggestion
## Wait a Bit
```

File: site/docs/tutorials/snake/setting-color-palette.md [hunk 25]
```diff
@@ -0,0 +1,128 @@
+---
+sidebar_label: Setting Color Palette
+---
+
+import MultiLanguageCode from '@site/src/components/MultiLanguageCode';
+
+# Setting Color Palette
+
+## The default color palette
```
Comment (aduros) [inline site/docs/tutorials/snake/setting-color-palette.md:9]:
```suggestion
## The Default Color Palette
```

File: site/docs/tutorials/snake/setting-color-palette.md [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+<div className="row row--no-gutters">
+    <div className="col col--2" style={{padding: "1.5rem", background: "#e0f8cf", color: "#000"}}>Color 1: #e0f8cf</div>
+    <div className="col col--2" style={{padding: "1.5rem", background: "#86c06c", color: "#000"}}>Color 2: #86c06c</div>
+    <div className="col col--2" style={{padding: "1.5rem", background: "#306850", color: "#fff"}}>Color 3: #306850</div>
+    <div className="col col--2" style={{padding: "1.5rem", background: "#071821", color: "#fff"}}>Color 4: #071821</div>
+</div>
+
+It's inspired by the color palette of the original gameboy. But you can set your own palette too.
+
+## Picking a color palette
```
Comment (aduros) [inline site/docs/tutorials/snake/setting-color-palette.md:22]:
```suggestion
## Picking a Color Palette
```

File: site/docs/tutorials/snake/setting-color-palette.md [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+In WASM-4, you can set 4 colors to your liking. The default color palette looks like this:
+
+<div className="row row--no-gutters">
+    <div className="col col--2" style={{padding: "1.5rem", background: "#e0f8cf", color: "#000"}}>Color 1: #e0f8cf</div>
+    <div className="col col--2" style={{padding: "1.5rem", background: "#86c06c", color: "#000"}}>Color 2: #86c06c</div>
+    <div className="col col--2" style={{padding: "1.5rem", background: "#306850", color: "#fff"}}>Color 3: #306850</div>
+    <div className="col col--2" style={{padding: "1.5rem", background: "#071821", color: "#fff"}}>Color 4: #071821</div>
+</div>
+
+It's inspired by the color palette of the original gameboy. But you can set your own palette too.
```
Comment (aduros) [inline site/docs/tutorials/snake/setting-color-palette.md:20]:
```suggestion
It's inspired by the color palette of the original Gameboy. But you can set your own palette too.
```

File: site/docs/tutorials/snake/setup-project.md [hunk 28]
```diff
@@ -0,0 +1,157 @@
+---
+sidebar_label: Setup your project
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+
+# Setup your project
```
Comment (aduros) [inline site/docs/tutorials/snake/setup-project.md:9]:
```suggestion
# Setup Your Project
```

File: site/docs/tutorials/snake/setup-project.md [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+
+Before you can begin, you need a project to work with. The CLI-Tool of WASM-4 includes templates for several programing languages like AssemblyScript, C, Go and Rust. More languages are sure to come in the future.
+
+## The Terminal ...
```
Comment (aduros) [inline site/docs/tutorials/snake/setup-project.md:13]:
```suggestion
## The Terminal
```

File: site/docs/tutorials/snake/setup-project.md [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+W4_NO_BROWSER=1 w4 watch
+```
+:::
+
+:::note Setting a default language
+You can set a default language by setting the environment variable `W4_LANG` to one of the supported languages.
+It's possible to override it with the CLI parameter later on.
+:::
+
+## Current state
```
Comment (aduros) [inline site/docs/tutorials/snake/setup-project.md:151]:
```suggestion
## Current State
```

File: site/docs/tutorials/snake/handling-user-input.md [hunk 31]
```diff
@@ -0,0 +1,330 @@
...
+| Button 1 | 1     |
+| Button 2 | 2     |
+| Left     | 16    |
+| Right    | 32    |
+| Up       | 64    |
+| Down     | 128   |
+
+So if the player presses "Right" and "Button 1", the value would be 33. Now are all values "a power of two", meaning you can set and check them using binary operators.
+
+If you wnt to check if Button 1 is pressed, simply use the binary AND:
```
Comment (aduros) [inline site/docs/tutorials/snake/handling-user-input.md:31]:
```suggestion
If you want to check if Button 1 is pressed, simply use the binary AND:
```

File: site/docs/tutorials/snake/handling-user-input.md [hunk 32]
```diff
@@ -0,0 +1,330 @@
+---
+sidebar_label: Handling user input
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Handling user input
+
+With a moving snake, you've done a good part of the game. This part let's you give the control into the players hands. If the player presses a certain button, you just have to change the values of the "Direction"-Property of your snake.
```
Comment (aduros) [inline site/docs/tutorials/snake/handling-user-input.md:10]:
```suggestion
With a moving snake, you've done a good part of the game. This part lets you give the control into the players hands. If the player presses a certain button, you just have to change the values of the "Direction"-Property of your snake.
```

**Review by aduros — COMMENTED** [review]
It looking great! I added a few suggestions, mainly minor stylistic things like converting the titles to [Title Case](https://en.wikipedia.org/wiki/Title_case) for consistency with the rest of the docs.

File: site/docs/tutorials/snake/moving-the-snake.md (hunk 2 above)
Comment (christopher-kleine) [reply site/docs/tutorials/snake/moving-the-snake.md:39]: I'm currently adding AssemblyScript.
Afterwards, I try my hands on Rust.
Then I try to get C working.
And I want them all finished before I'd remove the "Draft" status.

File: site/docs/tutorials/snake/handling-user-input.md [hunk 33]
```diff
@@ -0,0 +1,330 @@
...
+For the player side of things, WASM-4 tries to cover most keyboard-layouts:
+
+- X and Space = Button 1
+- Y, C and Z = Button 2
+
+This should cover QWERTY, QWERTZ and Dovrak layouts.
+
+The gamepads for the players 2, 3 and 4 are currently not implemented.
+
+## Detecting `just pressed`
```
[the remark this answers is not in the document]
Comment (christopher-kleine) [reply site/docs/tutorials/snake/handling-user-input.md:46]: I agree.

File: site/docs/tutorials/snake/moving-the-snake.md (hunk 1 above)
Comment (christopher-kleine) [reply site/docs/tutorials/snake/moving-the-snake.md:2]: I tried? I think. I'll check again and if it's not necessary, I'll remove it. I've seen your docs without them, so I was confused, why this didn't work for me.

But like I said: I'll check again.

File: site/docs/tutorials/snake/goal.md (hunk 3 above)
Comment (christopher-kleine) [reply site/docs/tutorials/snake/goal.md:27]: Yeah, I think it's important to know what you're going to create. And even more important: To know what you *won't* create.
People can get disappointed realy fast if they think they're building the next ultra-high FPS. ;-)

**Comment by aduros** [comment]
Thankssss :snake:

**Pull request merged**